python/lsst/ts/scheduler/model.py

```diff
@@ -523,6 +563,98 @@ async def callback_script_info(self, data: BaseDdsDataType) -> None:
             for key in items_to_remove:
                 del self.script_info[key]
 
+        if len(block_all_done) > self.max_scripts:
+            blocks_to_remove = block_all_done[: self.max_scripts - 1]
+            self.log.debug(f"Removing done blocks: {blocks_to_remove}")
+            for block_uid_done in blocks_to_remove:
+                del self._block_scripts_final_state[block_uid_done]
```

Comment:
Not sure that would be the case here though... In the dict comprehension case you suggest it would have to iterate over all the elements of the dictionary and discard the ones in `blocks_to_remove`. Then discard the whole dictionary and create a new one. 

In my code I am simply iterating over a smaller list of indices and deleting them from the dictionary. I did a small timing test to showcase it:

```
In [7]: def delete_items_from_dict():
   ...:     a_dict = {i: i for i in range(100000)}
   ...:     to_delete = [i for i in range(1000)]
   ...:     for i in to_delete:
   ...:         del a_dict[i]
   ...:

In [8]: def dict_comprehension():
   ...:     a_dict = {i: i for i in range(100000)}
   ...:     to_delete = [i for i in range(1000)]
   ...:     a_dict = {key: value for key, value in a_dict.items() if key not in to_delete}
   ...:

In [9]: %timeit delete_items_from_dict()
4.22 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: %timeit dict_comprehension()
853 ms ± 1.64 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

As you can see for large dicts and small delete list (which is what I expect this will look like), my version works much faster.

Reply:
Wow, thanks for the detailed analysis!